doc/source/dev/contributor/compiled_code.rst

```diff
@@ -8,24 +8,29 @@ Beyond Python
 
 This is a small collection of thoughts related to the inclusion of code written
 in languages other than Python. Currently, the only option for languages other
-than Python that is explicitly documented is :ref:`Cython<adding-cython>`.
+than Python that we have extra documentation for is :ref:`Cython<adding-cython>`.
 
 *Can I use a programming language other than Python to speed up my code?*
 
-Yes. The languages used in SciPy are Python, Cython, C, C++, and Fortran. All
-of these have their pros and cons. If Python really doesn't offer enough
-performance, one of those languages can be used. Important concerns when
+Yes. The languages used in SciPy are Python, Cython, Pythran, C, C++, and
+Fortran. All of these have their pros and cons. If Python really doesn't offer
+enough performance, one of those languages can be used. Important concerns when
 using compiled languages are maintainability and portability. For
-maintainability, Cython is clearly preferred over C/C++/Fortran. Cython and C
-are more portable than C++/Fortran. A lot of the existing C and Fortran code
-in SciPy is older, battle-tested code that was only wrapped in (but not
-specifically written for) Python/SciPy. Therefore, the basic advice is: use
-Cython. If there are specific reasons why C/C++/Fortran should be preferred,
+maintainability, Pythran and Cython are preferred over C/C++/Fortran. Cython, C
+and C++ are more portable than Fortran. A lot of the existing C and Fortran
+code in SciPy is older, battle-tested code that was only wrapped in (but not
+specifically written for) Python/SciPy.
+
+Our basic advice is: use Pythran or Cython for accelerating smaller pieces of
+code. In cases where Pythrand or Cython are no longer enough, prefer C++.
+If there are specific reasons why C and especially Fortran are preferred,
```

Comment:
Preferring C++ over plain C seems a bit more contentious. I'd say, well plain C isn't gruesome while C++ often is, so that's my reason for using it.

Not to mention that CPython is written in C.

Presumably there's some reasoning about supporting multiple types gracefully through templates, though often times `double` is pretty much good enough.

Reply:
I'm not sure why that would be contentious. We are adding new code in C++ regularly, including original code like `spatial.KDTree` and `fft` rather than only vendored code. We also have multiple devs that like working in C++. We have no examples in the last 5 years of writing in plain C I believe, only really old and hard to maintain code like `ndimage`.

For complex code where Cython is not enough or not the right fit, I think C++ is the best alternative nowadays.

Reply:
There are many high profile programmers who have been interviewed and stated that C can be beautiful while C++ is just an abomination. Peter Siebel's "Coders at Work: Reflections on the Craft of Programming " book is an excellent example of this.

So, fair enough, it seems this is the reality and we have devs who are keen on C++, but I don't think this is without contention (i.e., I don't want to read C++ and C is fair bit easier to get comfortable with). I thought Chuck also noted that at least a small part of the reasoning one might not want to rewrite NumPy in C++ instead of C was related to the skill needed to do C++ correctly and read/review it (i.e., accessibility).

Reply:
This is not a C vs. C++ argument. It's specifically: "when you want to write fast code for SciPy *and Cython doesn't cut it*, we'd prefer you to use X".

Reply:
Ok, we're not going to agree here--`X` is either `C` or `C++` and it looks like a decision between the two based on debatable preference to me. You can still use C for stuff Cython can't do. If it were up to me I'd probably go with other way, use C unless you need something fancy like templating, then that's a justification for C++.

Reply:
C++ is fine when written by experts and proven itself but for custom code I thlnk it's a hell to sift through the templates and other tricks in my opinion. Besides we have folks from numpy who can help us better for C but don't know if that's the case for c++

Reply:
Maybe we can just word it _C or C++_ and leave it at that. For the most part, the difference is going to be minor, I rather strongly suspect.

Unwanted stuff, like mallocs in the computational routines or fancy metaprogramming magic can be weeded out at the code review stage anyway. Going into code guidelines on how to write good C or what part of C++ we want is probably out of place in the scipy docs (or, if someone's up to formulating some guidelines, that's likely a separate document).

Reply:
> We have no examples in the last 5 years of writing in plain C I believe...

This PR contains plain C: https://github.com/scipy/scipy/pull/10170

Reply:
changed to "C or C++"